rfd/0005-kubernetes-service.md

```diff
@@ -0,0 +1,90 @@
+---
+authors: Andrew Lytvynov (andrew@gravitational.com)
+state: draft
+---
+
+# RFD 5 - Kubernetes service
+
+## What
+
+Dedicated Kubernetes service in a Teleport process.
+
+## Why
+
+Kubernetes integration in Teleport has been implemented as an optional addon to
+the Proxy service. This has several downsides:
+- doesn't follow the "separation of concerns" principle; proxy bundles in:
+  - public endpoint for Teleport users (both Web UI and CLI)
+  - Kubernetes API gateway
+  - reverse tunnel server for other services (like Nodes behind firewall/NAT or
+    trusted clusters)
+- pulls in all of the semantics of a Proxy service, like being a target for the
+  discovery protocol
+  - this makes exposing a Kubernetes API behind firewall much trickier
+
+The only benefit from bundling the Kubernetes access with the Proxy is direct
+access to the reverse tunnel server, for access to leaf Teleport clusters.
+
+The solution is to add a separate Kubernetes Service to the `teleport` binary.
+This fits with the existing model of Node service (SSH) and Application service
+([AAP](https://github.com/gravitational/teleport/issues/3951)).
+Proxy service becomes more of a "dumb" router in this model (not really "dumb",
+there's still a lot of authn/z and audit complexity).
+
+This RFD complements the [Kubernetes 5.0 enhancements
+design](https://docs.google.com/document/d/1cS6J2d_xBcJMWPewWPjdOZyrDHLKdqmgF1QLLo4E1YI).
+
+## Details
+
+### Service definition
+
+The new Kubernetes service will inherit all of the configuration and behavior
+of the existing `kubernetes` section of `proxy_service`:
+
+```yaml
+# Old format:
+proxy_service:
+    kubernetes:
+        enabled: yes
+        public_addr: [k8s.example.com:3026]
+        listen_addr: 0.0.0.0:3026
+        kubeconfig_file: /secrets/kubeconfig
+```
+
+but will exist in a new top-level config:
+
+```yaml
+# New format:
+kubernetes_service:
+    enabled: yes # default "no"
```

Comment:
With 5.0, if you recommend users set `Admin starts a Teleport agent within each Kubernetes cluster, with the following`, this would mean that the Auth/Proxy on the root could have this set to False? ( this would a similar to ssh_service being false ) so it would follow the same pattern.

Reply:
Admin can set it to false.
Proxy still needs to listed on a k8s port to forward requests though

Reply:
This is somewhat confusing - the idea that you could start a `kubernetes_service` and give it a `public_addr` but it still isn't actually the thing that's responsible for doing the listening. It doesn't work without a corresponding proxy, right?

Reply:
Ping @awly for comment here

Reply:
Sorry, missed this.
A `proxy_service` with `kube_listend_addr` set is a user-facing endpoint.
A `kubernetes_service` is a gateway for a single k8s cluster from a teleport cluster, that is only reachable through a proxy.

Some scenarios:
- single teleport cluster, single k8s cluster - use both `kubernetes_service` and `proxy_service` with `kube_listen_addr`
- single teleport cluster, multiple k8s clusters - `auth_service` and `proxy_service` on one box, separate pods with only `kubernetes_service` in each cluster
- root teleport cluster without local k8s cluster, leaf teleport clusters with k8s clusters - root proxy needs `kube_listen_addr` but **not** `kubernetes_service`; leaf proxies need `kube_listen_addr`, with `kubernetes_service` in one process or separate pods running just `kubernetes_service`

There's a consistent requirement - if you want k8s integration, your proxies must set `kube_listen_addr` and clients only talk to `kubernetes_service` through a proxy.